lib/segment/src/segment.rs

```diff
@@ -1112,86 +1144,98 @@ impl SegmentEntry for Segment {
             .collect()
     }
 
-    fn take_snapshot(&self, snapshot_dir_path: &Path) -> OperationResult<()> {
+    fn take_snapshot(&self, snapshot_dir_path: &Path) -> OperationResult<PathBuf> {
         log::debug!(
             "Taking snapshot of segment {:?} into {:?}",
             self.current_path,
             snapshot_dir_path
         );
+
         if !snapshot_dir_path.exists() {
-            return Err(OperationError::service_error(&format!(
-                "the snapshot path provided {:?} does not exist",
-                snapshot_dir_path
+            return Err(OperationError::service_error(format!(
+                "the snapshot path {snapshot_dir_path:?} does not exist"
             )));
         }
+
         if !snapshot_dir_path.is_dir() {
-            return Err(OperationError::service_error(&format!(
-                "the snapshot path provided {:?} is not a directory",
-                snapshot_dir_path
+            return Err(OperationError::service_error(format!(
+                "the snapshot path {snapshot_dir_path:?} is not a directory",
             )));
         }
+
         // flush segment to capture latest state
         self.flush(true)?;
-        // extract segment id from current path
+
+        let tmp_path = self.current_path.join("tmp");
+
+        let db_backup_path = tmp_path.join("db_backup");
+        let payload_index_db_backup_path = tmp_path.join("payload_index_db_backup");
+
+        {
+            let db = self.database.read();
+            crate::rocksdb_backup::create(&db, &db_backup_path)?;
+        }
+
+        self.payload_index
+            .borrow()
+            .take_database_snapshot(&payload_index_db_backup_path)?;
```

Comment:
It seems that with this approach, we could have inconsistent snapshots as vectors and payload are captured at different time.

Depending on how long `crate::rocksdb_backup::create(&db, &db_backup_path)` takes, updates could be applied on payload in the meantime.

Could this be a problem in real life?  Maybe documenting that collection should be left alone during snapshoting.

I am not sure the previous implementation was rock solid in that regard anyway.

WDYT?

Reply:
> It seems that with this approach, we could have inconsistent snapshots as vectors and payload are captured at different time.


The segment is locked for writing during the snapshot

Reply:
I was told that the Segment is write locked during the snasphot. My comment can be discarded!